molgenis-data-security/src/main/java/org/molgenis/data/security/audit/AuditTransactionListener.java

```diff
@@ -0,0 +1,38 @@
+package org.molgenis.data.security.audit;
+
+import static java.util.Objects.requireNonNull;
+import static org.molgenis.data.security.audit.AuthenticationUtils.getUsername;
+import static org.molgenis.data.security.audit.AuthenticationUtils.isRunByUser;
+
+import java.util.Map;
+import org.molgenis.audit.AuditEventPublisher;
+import org.molgenis.data.transaction.TransactionListener;
+import org.molgenis.data.transaction.TransactionManager;
+import org.springframework.stereotype.Component;
+
+/**
+ * Publishes audit events with transaction ids so that they can be correlated to other audit events
+ * that were published during that transaction.
+ */
+@Component
+public class AuditTransactionListener implements TransactionListener {
+
+  static final String TRANSACTION_FAILURE = "TRANSACTION_FAILURE";
+  static final String TRANSACTION_ID = "transactionId";
+
+  private final AuditEventPublisher auditEventPublisher;
+
+  public AuditTransactionListener(
+      TransactionManager transactionManager, AuditEventPublisher auditEventPublisher) {
+    requireNonNull(transactionManager).addTransactionListener(this);
+    this.auditEventPublisher = requireNonNull(auditEventPublisher);
+  }
+
+  @Override
+  public void rollbackTransaction(String transactionId) {
+    if (isRunByUser()) {
```

Comment:
In the case of jobs, not entirely sure if the context is still present at this moment

Reply:
I'll have a go at testing that

Reply:
For jobs a `SystemSecurityToken` is set I believe

Reply:
Depends on the job, when a job is run as a user, a token is set for the duration of the transaction. What I wonder is when exactly the token gets unset. Before or after the rollbackTransaction is called

Reply:
```{"timestamp":"2021-03-04T07:56:30.058490Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht36rlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.086132Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4ahlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.110494Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4bxlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.177179Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaae","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.177521Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaai","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.177796Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaam","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.178157Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaaq","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.178516Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaau","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.178880Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"it_emx_autoid_testAutoId","entityId":"aaaac574ht4f5lg6jaabciqaay","transactionId":"aaaac574ht4fjlg6jaabciqaaq"}}
{"timestamp":"2021-03-04T07:56:30.203098Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4hplg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.240295Z","principal":"admin","type":"ENTITY_CREATED","data":{"entityTypeId":"sys_FileMeta","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4iflg6jaabciqabe"}}
{"timestamp":"2021-03-04T07:56:30.284701Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4mvlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.294393Z","principal":"admin","type":"TRANSACTION_FAILURE","data":{"transactionId":"aaaac574ht4mvlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.308167Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4nvlg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:30.412463Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_FileIngestJobExecution","entityId":"aaaac574ht33zlg6jaabciqacu","transactionId":"aaaac574ht4uplg6jaabciqaae"}}
{"timestamp":"2021-03-04T07:56:42.509420Z","principal":"admin","type":"ENTITY_UPDATED","data":{"entityTypeId":"sys_job_ScheduledJob","entityId":"aaaac574gu2mplg6jaabciqaai","transactionId":"aaaac574huuh7lg6jaabciqaae"}}
```

Triggered by the scheduler. Looking good!